homeassistant/components/pi_hole/__init__.py

```diff
@@ -16,95 +16,192 @@
+
+    def ensure_api_token(call_data):
+        """Ensure the Pi-Hole to be enabled/disabled has a api_token configured."""
+        data = get_data()
+        if SERVICE_DISABLE_ATTR_NAME not in call_data:
+            for slug in data:
+                call_data[SERVICE_DISABLE_ATTR_NAME] = data[slug].name
+                ensure_api_token(call_data)
+
+            call_data[SERVICE_DISABLE_ATTR_NAME] = None
+        else:
+            slug = slugify(call_data[SERVICE_DISABLE_ATTR_NAME])
+
+            if (data[slug]).api.api_token is None:
+                raise vol.Invalid(
+                    "Pi-hole '{}' must have an api_key provided in configuration to be enabled.".format(
+                        pi_hole.name
+                    )
+                )
+
+        return call_data
+
+    SERVICE_DISABLE_SCHEMA = vol.Schema(  # pylint: disable=invalid-name
+        vol.All(
+            {
+                vol.Required(SERVICE_DISABLE_ATTR_DURATION): vol.All(
+                    cv.time_period_str, cv.positive_timedelta
+                ),
+                vol.Optional(SERVICE_DISABLE_ATTR_NAME): vol.In(
+                    list(map(lambda c: c[CONF_NAME], config[DOMAIN])),
+                    msg="Unknown Pi-Hole",
+                ),
+            },
+            ensure_api_token,
+        )
     )
 
-    await pi_hole.async_update()
-
-    hass.data[DOMAIN] = pi_hole
-
-    async def handle_disable(call):
-        if api_key is None:
-            raise vol.Invalid("Pi-hole api_key must be provided in configuration")
+    SERVICE_ENABLE_SCHEMA = vol.Schema(  # pylint: disable=invalid-name
+        {
+            vol.Optional(SERVICE_ENABLE_ATTR_NAME): vol.In(
+                list(map(lambda c: c[CONF_NAME], config[DOMAIN])), msg="Unknown Pi-Hole"
+            )
+        }
+    )
 
-        duration = call.data[SERVICE_DISABLE_ATTR_DURATION].total_seconds()
+    hass.data[DOMAIN] = {}
+
+    for conf in config[DOMAIN]:
+        name = conf[CONF_NAME]
+        host = conf[CONF_HOST]
+        use_tls = conf[CONF_SSL]
+        verify_tls = conf[CONF_VERIFY_SSL]
+        location = conf[CONF_LOCATION]
+        api_key = conf.get(CONF_API_KEY)
+
+        LOGGER.debug("Setting up %s integration with host %s", DOMAIN, host)
+
+        session = async_get_clientsession(hass, verify_tls)
+        pi_hole = PiHoleData(
+            Hole(
+                host,
+                hass.loop,
+                session,
+                location=location,
+                tls=use_tls,
+                api_token=api_key,
+            ),
+            name,
+        )
 
-        LOGGER.debug("Disabling %s %s for %d seconds", DOMAIN, host, duration)
-        await pi_hole.api.disable(duration)
+        await pi_hole.async_update()
 
-    async def handle_enable(call):
-        if api_key is None:
-            raise vol.Invalid("Pi-hole api_key must be provided in configuration")
+        hass.data[DOMAIN][slugify(name)] = pi_hole
```

Comment:
Slugifying the name must be done in the config schema. Otherwise we risk duplicates.

Reply:
I don't understand. This line is what saves the library object to `hass.data` so that the sensor entity can act on it.